truffle/src/main/java/org/jruby/truffle/core/encoding/EncodingManager.java

```diff
@@ -29,14 +29,16 @@
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
 import java.nio.charset.UnsupportedCharsetException;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 
 public class EncodingManager {
 
-    private final DynamicObject[] ENCODING_LIST_BY_ENCODING_LIST_INDEX = new DynamicObject[EncodingDB.getEncodings().size()];
-    private final DynamicObject[] ENCODING_LIST_BY_ENCODING_INDEX = new DynamicObject[EncodingDB.getEncodings().size()];
+    private final List<DynamicObject> ENCODING_LIST_BY_ENCODING_LIST_INDEX = new ArrayList<DynamicObject>(EncodingDB.getEncodings().size());
+    private final Map<Integer, DynamicObject> ENCODING_LIST_BY_ENCODING_INDEX = new HashMap<Integer,DynamicObject>(EncodingDB.getEncodings().size());
```

Comment:
Aliases resolve to their aliased encoding:

```
irb(main):002:0> 'abc'.encode('ASCII-8BIT').encoding.name
=> "ASCII-8BIT"
irb(main):003:0> 'abc'.encode('BINARY').encoding.name
=> "ASCII-8BIT"
```

Since we associate only the jcodings `Encoding` with the rope, we need to look up the associated `DynamicObject`. Prior to this change, there was essentially a 1:1 mapping, since aliases always resolved to their aliased encoding.

`Encoding#replicate`, however, works differently.

```
irb(main):004:0> 'abc'.encode('BINARY').encoding.name
=> "ASCII-8BIT"
irb(main):005:0> 'abc'.encode(Encoding::BINARY.replicate('blah')).encoding.name
=> "blah"
```

I'm pretty sure he made it a map so he could store multiple `DynamicObject`s with the same encoding index.